### DISCUSSION TITLE
Are constraints "strict" for Percival.jl?

### DISCUSSION THREAD

**Discussion by tpapp** [JuliaSmoothOptimizers/Organization#42] [Q&A]
If I have a problem with constraints `lvar .<= x .<= uvar`, is it guaranteed that the objective is always called with an `x` that satisfies these constraints in Percival.jl?

Empirically this seems not to be the case, I get occasional calls with `x` outside the domain. I am asking here before opening an issue (because isolating an MWE is tedious in my case, it is a rather large problem).

If the constraints are not "strict", what's the best strategy if eg `any(x .< lvar)` does not make sense? Returning a `Inf` objective?

**Comment by abelsiqueira** [comment]
Hi @tpapp, it is perfectly fine to ask here.
Percival.jl is an Augmented Lagrangian method whose subproblems are bound-constrained problems. It will penalize equalities but won't touch the bounds. The subproblems are currently solved with TRON, which I do believe enforces feasibility (numerical errors aside).
Inequalities are transformed into equalities by adding bounded slacks.

Are the calls far from the domain or are they close to numerical errors?
Are the bounds being input as bounds? If the mechanism for these bounds is actually creating inequalities, then the transformation could allow the initial variables go become infeasible (I think).

**Comment by tpapp** [comment]
@abelsiqueira, thanks for the super-quick answer.

The constraints are provided via the NLPModels interface in the `meta` variable as `lvar`, `uvar`. See [here](https://github.com/tpapp/ObjConsNLPModels.jl/blob/2acecc4623ce6d6d850f54f04a2d8209713cfb20/src/ObjConsNLPModels.jl#L109).

I get values which are wildly outside, not merely close to the boundary.

**Reply by abelsiqueira** [reply]
Then it is new territory. I tried a few short things to see if I could reproduce, but no luck.
Can you share some of the code that calls the optimizer?

**Reply by tpapp** [reply]
It comes from a 10k LOC problem, I will try to reduce it, worst case I share a self-contained private repo (would that be OK?).

**Reply by abelsiqueira** [reply]
That is ok

**Comment by tpapp** [comment]
Still working on an MWE, but I found that my original problem is singular, in the sense that locally there is a 1-dimensional manifold that is optimal and satisfies the constraints. Could that be causing a problem?

**Reply by abelsiqueira** [reply]
I don't see how - which doesn't mean that is not the issue. I can see the method failing/stalling, but not going out of bounds.
I tried something with `g(x) = (x[1] - 1)^2 + x[2]^2 - 4`, `f(y) = y^2`, `h(y) = [y]`, problem `min f(g(x)) s.to h(g(x)) = 0`.

Depending on the bounds and starting point it can converge to first-order or stall.
It can go out of bounds slightly - especially if I start out of bounds - but not a lot. How big is the infeasibility that you're seeing (absolute and relative values)?
